apacheconfig/wloader.py

```diff
@@ -31,19 +31,135 @@ def dump(self):
+
+    @classmethod
+    def parse(cls, raw_str, parser):
+        """Factory for :class:`apacheconfig.ListASTNode` from a config string.
+
+        Args:
+            raw_str (str): Config string to parse.
+            parser (:class:`apacheconfig.ApacheConfigParser`): parser object
+                to use.
+        Returns:
+            :class:`apacheconfig.ListASTNode` containing data parsed from
+            ``raw_str``.
         """
+        raw = parser.parse(raw_str)
+        return cls(raw, parser)
 
-    @abc.abstractmethod
-    @whitespace.setter
-    def whitespace(self, value):
-        """Set preceding or trailing whitespace for this node.
+    def add(self, index, raw_str):
+        """Parses and adds child element at given index.
+
+        Parses given string into an ASTNode object, then adds to list at
+        specified index.
 
         Args:
-            value (str): value to set whitespace to.
+            index (int): index of list at which to insert the node.
+            raw_str (str): string to parse. The parser will automatically
+                determine whether it's a :class:`apacheconfig.BlockNode` or
+                :class:`apacheconfig.LeafASTNode`.
+
+        Returns:
+            The :class:`apacheconfig.Node` created from parsing ``raw_str``.
         """
+        raw = self._parser.parse(raw_str)[1]
+        if raw[0] == "block":
+            node = BlockNode(raw, self._parser)
+        else:
+            node = LeafASTNode(raw)
+
+        # If we're adding an element to the beginning of list, the first
+        # item may not have a preceding newline-- so we add one in case.
+        # For instance, something like:
+        #   Contents("line1\nline2").add(0, "\nline0")
+        # should end up as "\nline0\nline1\nline2"
+        if (len(self._contents) >= 1 and index == 0 and
+           '\n' not in self._contents[0].whitespace):
+            whitespace_after = self._contents[0].whitespace
+            self._contents[0].whitespace = '\n' + whitespace_after
+        self._contents.insert(index, node)
+        return node
+
+    def remove(self, index):
+        """Removes node from supplied index.
+
+        Args:
+            index (int): index of node to remove from list.
+
+        Returns:
+            The :class:`apacheconfig.Node` that was removed.
+        """
+        thing = self._contents[index]
+        del self._contents[index]
+        return thing
+
+    def __len__(self):
+        """Number of :class:`apacheconfig.ASTNode` children."""
+        return len(self._contents)
+
+    def __iter__(self):
+        """Iterator over :class:`apacheconfig.ASTNode` children."""
+        return iter(self._contents)
+
+    def dump(self):
+        return ("".join([item.dump() for item in self._contents])
+                + self.trailing_whitespace)
+
+    @property
+    def typestring(self):
+        """See base class.
+
+        Can only return ``"contents"`` for :class:`apacheconfig.ListNode`.
```

Comment:
nit: here and at many other docstrings - may be we should out such explanations into formal `Returns:` section?